### PR TITLE
[PTDT-2863]: Feature schema attributes

### PR DESCRIPTION
**Pull request by Tim-Kerr**
# Description

Add feature schema attribute support to the `Tool` and `Classification` objects in the sdk.

The sdk now supports feature schema attributes either using the `OntologyBuilder` or by passing in raw JSON containing the attributes.

Test the following endpoints in the client:

* get_feature_schema
* get_feature_schemas
* create_ontology_from_feature_schemas
* delete_unused_feature_schema
* update_feature_schema_title
* upsert_feature_schema
* create_feature_schema

Test the creation of ontologies using the ontology builder with nodes containing feature schema attributes

Testing script:
```
import os
from labelbox import Client
from labelbox.schema.ontology import OntologyBuilder, Tool, PromptIssueTool
from labelbox.schema.tool_building.classification import Classification, Option
from labelbox.schema.tool_building.types import FeatureSchemaAttribute
from labelbox.schema.media_type import MediaType
from labelbox.schema.ontology_kind import OntologyKind


client = Client(
    api_key=os.environ.get('STAGE_API_KEY'),
    endpoint="https://app.lb-stage.xyz/api/_gql/graphql",
    rest_endpoint="https://app.lb-stage.xyz/api/api/v1")

builder = OntologyBuilder(

  tools=[
    Tool(
      name="Auto OCR",
      tool=Tool.Type.BBOX,
      attributes=[
        FeatureSchemaAttribute(
          attributeName="auto-ocr",
          attributeValue="true"
        )
      ],
      classifications=[
        Classification(
          name="Auto ocr text class value",
          instructions="This is an auto OCR text value classification",
          class_type=Classification.Type.TEXT,
          scope=Classification.Scope.GLOBAL,
          attributes=[
            FeatureSchemaAttribute(
              attributeName="auto-ocr-text-value",
              attributeValue="true"
            )
          ]
        )
      ]
    )
  ]
)

# client.create_ontology("Auto OCR ontology from sdk", builder.asdict(), media_type=MediaType.Document)

builder = OntologyBuilder(
  classifications=[
    Classification(
      name="prompt message scope text classification",
      instructions="This is a prompt message scoped text classification",
      class_type=Classification.Type.TEXT,
      scope=Classification.Scope.INDEX,
      attributes=[
        FeatureSchemaAttribute(
          attributeName="prompt-message-scope",
          attributeValue="true"
        )
      ]
    )
  ]
)

# client.create_ontology('MMC Ontology with prompt message scope class', builder.asdict(), media_type=MediaType.Conversational, ontology_kind=OntologyKind.ModelEvaluation)

builder = OntologyBuilder(
  classifications=[
    Classification(
      name="Requires connection checklist classification",
      instructions="This is a requires connection checklist classification",
      class_type=Classification.Type.CHECKLIST,
      scope=Classification.Scope.GLOBAL,
      attributes=[
        FeatureSchemaAttribute(
          attributeName="requires-connection",
          attributeValue="true"
        )
      ],
      options=[
        Option(
          value='First option'
        ),
        Option(
          value='Second option'
        )
      ]
    )
  ]
);

# client.create_ontology('Image ontology with requires connection classes', builder.asdict(), media_type=MediaType.Image)


# feature_schema = client.upsert_feature_schema(Tool(name='Testing', tool=Tool.Type.BBOX, attributes=[FeatureSchemaAttribute(attributeName='auto-ocr', attributeValue='true')]).asdict())
# print(feature_schema)
# fetched_feature_schema = client.get_feature_schema(feature_schema.uid)
# feature_schemas_with_name = client.get_feature_schemas('Auto OCR')

# # Iterate over the feature schemas
# for schema in feature_schemas_with_name:
#     print(schema)

# ontology = client.create_ontology_from_feature_schemas('Ontology from feature schemas', ['cm4rc1nl90h36070782v9hlpt'])

# feature_schema = client.update_feature_schema_title('cm4rc1nl90h36070782v9hlpt', 'This is a new title - did it remove the feature schema attributes? UPDATED')
# client.delete_unused_feature_schema('cm4rhzhn7026e07wm2az681di')

# feature_schema = client.create_feature_schema(normalized={'tool': 'rectangle',  'name': 'cat', 'color': 'black', 'attributes': [{'attributeName': 'auto-ocr', 'attributeValue': 'true'}]})
# print(feature_schema)

# classification = Classification.from_dict({
#     "name": "Test Classification",
#     "instructions": "Test instructions",
#     "type": "text",  # or "checklist" or other valid Classification.Type values
#     "scope": "index",  # or "index" for Classification.Scope values
#     "required": False,  # optional
#     "attributes": [  # optional
#         {
#             "attributeName": "prompt-message-scope",
#             "attributeValue": "true"
#         }
#     ],
#     "options": []
# })

# tool = Tool.from_dict({
#     "name": "Test Tool",
#     "type": "rectangle",  # or "checklist" or other valid Classification.Type values
#     "required": False,  # optional
#     "attributes": [  # optional
#         {
#             "attributeName": "auto-ocr",
#             "attributeValue": "true"
#         }
#     ],
#     "options": []
# })
```

Please include a summary of the changes and the related issue. Please also include relevant motivation and context.

Fixes # (issue)

## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Document change (fix typo or modifying any markdown files, code comments or anything in the examples folder only)

## All Submissions

- [x] Have you followed the guidelines in our Contributing document?
- [x] Have you provided a description?
- [x] Are your changes properly formatted?

## New Feature Submissions

- [x] Does your submission pass tests?
- [x] Have you added thorough tests for your new feature?
- [ ] Have you commented your code, particularly in hard-to-understand areas?
- [x] Have you added a Docstring?

## Changes to Core Features

- [x] Have you written new tests for your core changes, as applicable?
- [x] Have you successfully run tests with your changes locally?
- [x] Have you updated any code comments, as applicable?

### CODE REVIEW

File: libs/labelbox/src/labelbox/schema/ontology.py [hunk 1]
```diff
@@ -73,6 +74,7 @@ class Tool:
         classifications: (list)
         schema_id: (str)
         feature_schema_id: (str)
+        attributes: (list)
     """
```
Comment (Tim-Kerr) [inline libs/labelbox/src/labelbox/schema/ontology.py:82]: I noticed this Tool class doesn't reference a __post_init() method, which is where I print the attribute experimental warning. Should I add one?

Comment (vbrodsky) [reply libs/labelbox/src/labelbox/schema/ontology.py:82]: yes

File: libs/labelbox/src/labelbox/schema/tool_building/classification.py [hunk 2]
```diff
@@ -5,7 +5,10 @@
 
 from lbox.exceptions import InconsistentOntologyException
 
-from labelbox.schema.tool_building.types import FeatureSchemaId
+from labelbox.schema.tool_building.types import (
+    FeatureSchemaId,
```
Comment (vbrodsky) [inline libs/labelbox/src/labelbox/schema/tool_building/classification.py:9]: is this used?

File: libs/labelbox/src/labelbox/schema/tool_building/types.py [hunk 3]
```diff
@@ -1,6 +1,17 @@
-from typing import Annotated
-
+from typing import Annotated, List
 from pydantic import Field
+from typing import TypedDict
+
+
+class FeatureSchemaAttribute(TypedDict):
```
Comment (vbrodsky) [inline libs/labelbox/src/labelbox/schema/tool_building/types.py]: Can we make this into a data class? We already use Pydantic and dataclasses, so adding another paradigm feels unnecessary. Also, I’ve found TypedDict has quirks that make it less reliable. Personally, I treat it as a temporary internal type.

File: libs/labelbox/src/labelbox/schema/tool_building/types.py [hunk 4]
```diff
@@ -1,6 +1,17 @@
...
+    attributeName: str
+    attributeValue: str
+
+
+FeatureSchemaAttriubte = Annotated[FeatureSchemaAttribute, Field()]
```
Comment (vbrodsky) [inline libs/labelbox/src/labelbox/schema/tool_building/types.py]: FeatureSchemaAttriubte - typo?

File: libs/labelbox/src/labelbox/schema/tool_building/types.py [hunk 5]
```diff
@@ -1,6 +1,17 @@
...
 
 FeatureSchemaId = Annotated[str, Field(min_length=25, max_length=25)]
 SchemaId = Annotated[str, Field(min_length=25, max_length=25)]
+FeatureSchemaAttributes = Annotated[
```
Comment (vbrodsky) [inline libs/labelbox/src/labelbox/schema/tool_building/types.py:31]: good stuff

File: libs/labelbox/src/labelbox/schema/ontology.py (hunk 1 above)
Comment (Tim-Kerr) [reply libs/labelbox/src/labelbox/schema/ontology.py:82]: Added 👍

File: libs/labelbox/src/labelbox/schema/tool_building/classification.py (hunk 2 above)
Comment (Tim-Kerr) [reply libs/labelbox/src/labelbox/schema/tool_building/classification.py:9]: It's referenced [here](https://github.com/Labelbox/labelbox-python/pull/1930/files#diff-82372837b82fb0c5c38881008d0d346d23b6448f947af8513f69a42ca7991246R180)

File: libs/labelbox/src/labelbox/schema/tool_building/types.py (hunk 4 above)
Comment (Tim-Kerr) [reply libs/labelbox/src/labelbox/schema/tool_building/types.py]: Yes ty, fixed

File: libs/labelbox/src/labelbox/schema/tool_building/types.py (hunk 3 above)
Comment (Tim-Kerr) [reply libs/labelbox/src/labelbox/schema/tool_building/types.py]: Done 👍

File: libs/labelbox/src/labelbox/schema/ontology.py [hunk 6]
```diff
@@ -95,6 +101,13 @@ class Type(Enum):
     classifications: List[Classification] = field(default_factory=list)
     schema_id: Optional[str] = None
     feature_schema_id: Optional[str] = None
+    attributes: Optional[FeatureSchemaAttributes] = None
+
+    def __post_init__(self):
+        if self.attributes is not None:
+            warnings.warn(
+                "Attributes are an experimental feature and may change in the future."
```
Comment (vbrodsky) [inline libs/labelbox/src/labelbox/schema/ontology.py]: can you use a std message as described here (for beta) https://labelbox.atlassian.net/wiki/spaces/ENG/pages/3178037250/SDK+Experimental+Features

**Review by vbrodsky — DISMISSED** [review]

**Review by vbrodsky — APPROVED** [review]

**Pull request merged**